src/main/java/com/streamr/client/Subscription.java

```diff
@@ -174,4 +192,39 @@ void resendDone() {
 
         this.handler.done(this);
     }
+
+    //TODO: This method returns a new StreamMessage because MessageHandler takes a StreamMessage as argument.
```

Comment:
Maybe the problem is that this code might be in the wrong place? It doesn't feel like this should be in `Subscription`. 

Perhaps it belongs to `StreamMessage`? The decryption key could be passed to it as needed via a nice API. By symmetry the API could also include methods for encryption? Would be also good to reuse the stuff in `StreamMessage`, in particular we should avoid repeating something like `HttpUtils.mapAdapter.fromJson(content)` anywhere else than in `StreamMessage#getContent()`.

I'm just throwing quick ideas, but one solution with a nice developer-facing API would be to make the message stateful when in comes to encryption/decryption. This would make it easy to pass an "already decrypted" message to the handler. Some examples of usage:

```
// Maybe getContent() throws if the message is encrypted and hasn't been decrypted?
msg.getContent();

// "Unlock" the message
if (msg.isEncrypted()) {
    msg.decrypt(key);
}

// getContent() would now succeed without throwing
msg.getContent();

// Maybe there's a method to get the new key in case of NEW_KEY_AND_AES
// Simply returns null if this message contains no new key
msg.getNewGroupKey();
```

I guess encryption could work similarly:

```
// Create message normally
new StreamMessageV31(...)

// Encrypt in place
msg.encrypt(key);
```

Beauty of symmetry:
```
String originalMsg = msg.toJson();
msg.decrypt(key);
msg.encrypt(key);
assert originalMsg.equals(msg.toJson());
```

Reply:
I went with a slightly different approach and put the encryption/decryption methods in `EncryptionUtil` instead of `StreamMessage`. This is to have a clear separation of concern: the protocol files define only the structure of the exchanged messages, the logic of how to handle them is defined outside. This is the same approach we used for signing: everything is defined in `SigningUtil` not in `StreamMessage`.